### PR TITLE
Added possibility to have multiple test suites.

### PR DESCRIPTION
**Pull request by jpakkane**
This adds a `suite` keyword argument to tests. The Ninja backend also generates top level test targets for each available suite. As an example if you specify a test with `suite : 'slowsuite'` then you can run only tests in that suite with `ninja test-slowsuite`.

### CODE REVIEW

**Comment by rhd** [comment]
This looks good - I'll give it a test today. Can a test be part of multiple suites? `suite: ['cuda', 'slow']`

**Comment by jpakkane** [comment]
No, only one. But you can simulate that with `ninja test-cuda && ninja test-slow`.

**Comment by rhd** [comment]
Just tested it... looks really good. I'm happy to see this works:

```
test('t1', runtest,
    suite: 'slow',
)

test('t1', runtest,
    suite: 'fast',
)
```

**Comment by rhd** [comment]
After more poking around...

```
test('t1', runtest,
    suite: 'slow',
)

test('t1', runtest,
    suite: 'fast',
)
```

`ninja test` runs both of these tests. Since it's the same test, it would be really great if they were consolidated somehow.

File: meson_test.py [hunk 1]
```diff
@@ -144,6 +146,11 @@ def drain_futures(futures):
         (result, numlen, tests, name, i, logfile, jsonlogfile) = i
         print_stats(numlen, tests, name, result.result(), i, logfile, jsonlogfile)
 
+def filter_tests(suite, tests):
+    if suite is None:
+        return tests
```
Comment (rhd) [inline meson_test.py]: This should consolidate tests that are the same but have different 'suite' names

**Comment by jpakkane** [comment]
Getting into deduplication business is usually a neverending swamp. It's probably better to make the change discussed above, where one test can be part of multiple suites.

**Comment by rhd** [comment]
I'm a strong +1 on a single test can be in one or more test suites. I'll be happy to give it a test when it's ready.

**Comment by jpakkane** [comment]
Updated.

**Comment by rhd** [comment]
This change has worked pretty perfectly for me - thanks a lot.

**Comment by rhd** [comment]
The test count isn't being filtered. It says 1/40. It should be 1/1.

```
$ ninja -C build-gcc-suite-test/ test-runTest1
ninja: Entering directory `build-gcc-suite-test/'
1/1] Running test suite runTest1.
 1/40 runTest1 / backproject runTest1         OK       3.11 s

Full log written to meson-logs/testlog.txt.

```

**Comment by rhd** [comment]
Hi Jussi,

This is great and is working perfectly.

Just my opinion...  when a user types `ninja test`, they shouldn't see any suite names (you're call - this is a small issue). More then likely, each test for me will be in 1) it's own group based on the testname and 2) other larger groups ex: `opencl-slow` and `opencl`.

```
$ ninja -C build-gcc-suite-test/ test
ninja: Entering directory `build-gcc-suite-test/
[1/1] Running all tests.
 1/39 test1 / test1                           OK       3.13 s
 2/39 opencl / test2                          OK       2.86 s
```
vs
```
$ ninja -C build-gcc-suite-test/ test
ninja: Entering directory `build-gcc-suite-test/
[1/1] Running all tests.
 1/39 test1                            OK       3.13 s
 2/39 test2                            OK       2.86 s
```

Thanks a lot for this addition.

**Comment by rhd** [comment]
Hi Jussi, I'm really glad this landed! I just wanted to make sure you saw my previous update:

> Basically what I'm trying to say is that unless you specify the test suite name [on the command line], you shouldn't see it. Printing out `suite[0]` is arbitrary and could be really misleading. Consider this: suite: ['windows', 'linux']. You run `ninja test` on a linux box and it prints out `windows / test1`.

I believe 'ninja test' should not print out `suite[0]`

I'll live with whatever you decide - I just wanted to make sure you saw this. 

I'm amazed at how great you are making meson - thanks a lot.

**Comment by jpakkane** [comment]
It needs to print out the suite to clearly separate subprojects. Anyhow, that is only text, we can change it without any backwards compatibility problems later once we see how people are using it.

**Comment by rhd** [comment]
Oh, ok. I'm not using subprojects. :+1:

**Pull request merged**